File: server/engine/npc.go
package engine

import (
	"math/rand/v2"

	"kingdoms/engine/assets"
)

type npc struct {
	id             uint32
	typeID         uint8
	position       Position
	origin         Position
	health         float32
	targetID       uint32
	hasTarget      bool
	looking        bool
	destination    Position
	hasDestination bool
	movement       string
	mode           uint8
	usedModes      []bool
	modeTimer      float32
	attack         uint8
	attackTimer    float32
	damage         map[uint32]float32
	dead           bool
}

func newNPC(id uint32, typeID uint8, position Position, data assets.NPC) *npc {
	return &npc{
		id:        id,
		typeID:    typeID,
		position:  position,
		origin:    position,
		health:    data.Health,
		movement:  "wander",
		usedModes: make([]bool, len(data.Modes)),
		damage:    make(map[uint32]float32),
	}
}

func (n *npc) state() NPCState {
	return NPCState{
		ID:       n.id,
		Type:     n.typeID,
		X:        n.position.X,
		Y:        n.position.Y,
		Health:   n.health,
		TargetID: n.targetID,
		Targeted: n.looking && n.hasTarget,
	}
}

func (n *npc) updateTarget(characters map[uint32]*character, data assets.NPC) {
	maxDistance := min(data.Range, renderDistance)
	nearestDistance := maxDistance
	var nearestID uint32
	found := false

	for id, character := range characters {
		if character.dead {
			continue
		}
		currentDistance := distance(n.position, character.position)
		if currentDistance < nearestDistance {
			nearestDistance = currentDistance
			nearestID = id
			found = true
		}
	}

	n.targetID = nearestID
	n.hasTarget = found
	if !found {
		n.looking = false
	}
}

func (n *npc) tick(seconds float32, data assets.NPC, target *character, catalog *assets.Catalog) bool {
	n.modeTimer -= seconds
	n.attackTimer -= seconds

	if !n.hasTarget || target == nil || target.dead {
		n.hasTarget = false
		n.looking = false
		n.modeTimer = 0
		return false
	}

	if !n.validMode(data, n.mode) || n.modeTimer <= 0 {
		if int(n.mode) < len(n.usedModes) {
			n.usedModes[n.mode] = true
		}
		n.newMode(data)
	}

	if n.movement == "hover" && !n.hovering(data, target.position) {
		return false
	}

	return n.attackTimer < 0 && n.canAttack(data, target.position, catalog)
}

func (n *npc) validMode(data assets.NPC, index uint8) bool {
	if int(index) >= len(data.Modes) {
		return false
	}
	mode := data.Modes[index]
	if mode.SingleUse && n.usedModes[index] {
		return false
	}
	if mode.MaxHealth < n.health {
		return false
	}
	return mode.MinHealth < n.health
}

func (n *npc) newMode(data assets.NPC) {
	n.looking = false
	modes := make([]uint8, 0, len(data.Modes))

	for index, mode := range data.Modes {
		if !n.validMode(data, uint8(index)) {
			continue
		}
		modes = append(modes, uint8(index))
		if mode.Priority {
			modes = modes[len(modes)-1:]
			break
		}
	}

	if len(modes) == 0 {
		return
	}

	n.mode = modes[rand.IntN(len(modes))]
	mode := data.Modes[n.mode]
	n.modeTimer = mode.Duration
	n.movement = mode.Movement
	n.attackTimer = 0
}

func (n *npc) currentAttack(data assets.NPC) (assets.Attack, bool) {
	if int(n.mode) >= len(data.Modes) {
		return assets.Attack{}, false
	}
	attacks := data.Modes[n.mode].Attacks
	if len(attacks) == 0 {
		return assets.Attack{}, false
	}
	return attacks[int(n.attack)%len(attacks)], true
}

func (n *npc) canAttack(data assets.NPC, target Position, catalog *assets.Catalog) bool {
	attack, valid := n.currentAttack(data)
	if !valid {
		return false
	}

	var attackRange float32
	for _, spawn := range attack.Projectiles {
		projectile, _ := catalog.Projectile(spawn.ID)
		attackRange = max(attackRange, projectile.Range)
	}
	if len(attack.Bombs) > 0 {
		attackRange = max(attackRange, 32)
	}
	if len(attack.Summons) > 0 {
		attackRange = max(attackRange, data.Range)
	}
	return distance(n.position, target) < attackRange
}

File: training/src/kingdoms_sim/npc.cpp
#include "kingdoms_sim/npc.hpp"

#include <algorithm>

namespace kingdoms {

bool Npc::ValidMode(const AssetDatabase& assets, uint8_t idx) const {
  const NpcModeData& mode = assets.Npc(type_id).modes[idx];
  if (mode.single_use && used_modes[idx]) return false;
  if (mode.max_health < health) return false;
  if (mode.min_health >= health) return false;
  return true;
}

void Npc::NewMode(const AssetDatabase& assets, std::mt19937& rng) {
  const NpcData& data = assets.Npc(type_id);
  std::vector<uint8_t> pool;

  for (uint8_t idx = 0; idx < data.modes.size(); ++idx) {
    if (ValidMode(assets, idx)) {
      pool.push_back(idx);
      if (data.modes[idx].priority) {
        pool = {idx};
        break;
      }
    }
  }

  if (!pool.empty()) {
    std::uniform_int_distribution<size_t> dist(0, pool.size() - 1);
    uint8_t chosen = pool[dist(rng)];
    mode = chosen;
    mode_timer = data.modes[chosen].duration;
    movement = data.modes[chosen].movement;
    attack_timer = 0.0f;
  }
}

const AttackData* Npc::CurrentAttack(const AssetDatabase& assets) const {
  const NpcModeData& m = assets.Npc(type_id).modes[mode];
  if (m.attacks.empty()) return nullptr;
  return &m.attacks[attack % m.attacks.size()];
}

bool Npc::CanAttack(const AssetDatabase& assets, Vec2 target_pos) const {
  if (!InCombat()) return false;
  const AttackData* atk = CurrentAttack(assets);
  if (!atk) return false;

  float attack_range = 0.0f;
  for (const auto& p : atk->projectiles) {
    attack_range = std::max(attack_range, assets.Projectile(p.id).range);
  }
  if (!atk->bombs.empty()) attack_range = 32.0f;
  if (!atk->summons.empty()) attack_range = std::max(attack_range, assets.Npc(type_id).range);

  return Distance(pos, target_pos) < attack_range;
}

Npc MakeNpc(const AssetDatabase& assets, uint8_t type_id, Vec2 pos) {
  Npc npc;
  npc.type_id = type_id;
  npc.pos = pos;
  npc.origin = pos;
  npc.movement_point = pos;  // forces Wander to pick a fresh roam point on its first tick
  npc.health = assets.Npc(type_id).health;
  npc.used_modes.assign(assets.Npc(type_id).modes.size(), false);
  return npc;
}

}  // namespace kingdoms
